File: json-schema/research/format_uri/native_probes/native.mjs
// JS NATIVE URI-parser probe (WHATWG URL, Node). For each input, reports what
// `new URL(value)` thinks. WHATWG URL requires an absolute URL (a base-less
// URL must be absolute), so successful construction ~ "valid absolute URL".
// WHATWG URL is famously LENIENT and NORMALIZING (lowercases scheme/host,
// rewrites backslashes for special schemes, percent-encodes, etc).
//
// Emits JSON Lines: {"id","engine":"js-native","valid":bool,"detail":string}
// Run: node native.mjs ../native_inputs.json
import { readFileSync } from "node:fs";

const path = process.argv[2] ?? "../native_inputs.json";
const corpus = JSON.parse(readFileSync(path, "utf8"));

for (const inp of corpus.inputs) {
  let valid = false;
  let detail = "";
  try {
    const u = new URL(inp.value);
    valid = true;
    detail = `href=${u.href}`; // shows normalization
  } catch (e) {
    valid = false;
    detail = `error: ${e.message}`;
  }
  process.stdout.write(
    JSON.stringify({ id: inp.id, engine: "js-native", valid, detail }) + "\n",
  );
}

File: specs/json-schema/research/format_uri/native_probes/native.go
// Go NATIVE URI-parser probe. For each input, reports what Go's net/url thinks:
//   - url.Parse succeeds AND result.IsAbs() (has a scheme) => "valid absolute"
// This mirrors what a naive "use the stdlib" format:uri validator would do.
//
// Emits JSON Lines: {"id","engine":"go-native","valid":bool,"detail":string}
// Run: go run native.go ../native_inputs.json
package main

import (
	"encoding/json"
	"fmt"
	"net/url"
	"os"
)

type input struct {
	ID    string `json:"id"`
	Value string `json:"value"`
}
type corpus struct {
	Inputs []input `json:"inputs"`
}

func main() {
	path := "../native_inputs.json"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}
	data, err := os.ReadFile(path)
	if err != nil {
		fmt.Fprintln(os.Stderr, "read error:", err)
		os.Exit(1)
	}
	var c corpus
	if err := json.Unmarshal(data, &c); err != nil {
		fmt.Fprintln(os.Stderr, "parse error:", err)
		os.Exit(1)
	}
	enc := json.NewEncoder(os.Stdout)
	for _, in := range c.Inputs {
		valid := false
		detail := ""
		u, perr := url.Parse(in.Value)
		if perr != nil {
			detail = "parse-error: " + perr.Error()
		} else if !u.IsAbs() {
			detail = "not-absolute (no scheme)"
		} else {
			valid = true
			detail = "scheme=" + u.Scheme
		}
		_ = enc.Encode(map[string]any{
			"id": in.ID, "engine": "go-native", "valid": valid, "detail": detail,
		})
	}
}
